src/client/opamSwitchCommand.ml

```diff
@@ -494,18 +494,65 @@ let import_t ?ask importfile t =
   end;
   t
 
-let export rt ?(full=false) filename =
-  let switch = OpamStateConfig.get_switch () in
+let freeze_opam src_dir nv opam =
+  match OpamFile.OPAM.url opam with
+  | None -> opam
+  | Some url ->
+    let url_t = OpamFile.URL.url url in
+    match url_t.backend with
+    | #OpamUrl.version_control ->
+      (match OpamProcess.Job.run
+               (OpamRepository.revision (src_dir nv) url_t) with
+      | None ->
+        OpamConsole.error_and_exit `Not_found
+          "Unable to retrieve %s url revision, \
+           it can't be exported with --freeze."
+          (OpamPackage.to_string nv)
+      | Some hash ->
+        let opam =
+          OpamFile.OPAM.with_url
+            (OpamFile.URL.with_url
+               { url_t with hash = Some (OpamPackage.Version.to_string hash) }
+               url)
+            opam
+        in
+        opam)
+    | `http ->
+      (match OpamFile.URL.checksum url with
+       | [] ->
+         OpamConsole.error_and_exit `Not_found
+           "%s url doesn't have an associated checksum, \
+            it can't be exported with --freeze."
```

Comment:
sure, that is an option. I did not write that code, since I'm unsure how this should look like -- once here, can it be assumed that the sources are already downloaded (I suspect no), thus export would now do potentially lead to network communication.

Reply:
If the cache is not cleaned, the archive can be no more locally reachable, and needs a redownload.
On this part, I made it referring to https://github.com/ocaml/opam/pull/4055#discussion_r388347604